### PR TITLE
[SPARK-43124][SQL] Dataset.show projects CommandResults locally

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
`DataSet.show()` currently triggers a job for a simple `show tables` command. This is because the command output contains an `isTemporary` boolean column that needs to be casted to string when we use `show()` on the dataset.
This PR converts `CommandResult` to `LocalRelation` and let `ConvertToLocalRelation` to do the casting locally to avoid triggering job execution.

### Why are the changes needed?
A simple `show tables` shouldn not require an executor.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added new UT.

### CODE REVIEW

**Comment by peter-toth** [comment]
cc @cloud-fan, please let me know if you have a better idea.

**Comment by cloud-fan** [comment]
shall we update `ConvertToLocalRelation` to support `CommandResult` as well?

**Comment by peter-toth** [comment]
> or a more surgical way is to update `Dataset.getRows`, convert `CommandResult` to `LocalRelation` in the logical plan, and then execute the new logical plan.

Thanks, that's a good idea!

**Comment by peter-toth** [comment]
Updated to `LocalRelation` conversion in `Dataset.getRows()` in https://github.com/apache/spark/pull/40779/commits/98bf071cfb016b6461545a964ec07576419a6bb6

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -272,16 +272,32 @@ class Dataset[T] private[sql](
...
     val newDf = toDF()
-    val castCols = newDf.logicalPlan.output.map { col =>
-      // Since binary types in top-level schema fields have a specific format to print,
-      // so we do not cast them to strings here.
-      if (col.dataType == BinaryType) {
-        Column(col)
-      } else {
-        Column(col).cast(StringType)
-      }
+    val data = newDf.logicalPlan match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:275]: to deduplicate code, how about
```
val newDf = logicalPlan match {
  case c: CommandResult =>
    Dataset.ofRows(sparkSession, LocalRelation(c.output, c.rows)
  case _ => toDf()
}
same code as before...
```

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:275]: if people exclude `ConvertToLocalRelation`, then we shouldn't do local evaluation here as well.

Comment (peter-toth) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: All right, then https://github.com/apache/spark/pull/40779/commits/d8210ff849727961681b7d9df71b70e19ae24d41 contains the deduplication. Thanks for the idea.

**Review by HyukjinKwon — APPROVED** [review]

**Review by roryqi — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by peter-toth** [comment]
Thanks all for the review!

**Pull request closed without merging**